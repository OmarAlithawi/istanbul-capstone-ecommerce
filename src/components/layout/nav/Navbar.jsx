import React, { useReducer } from "react";
import Container from "react-bootstrap/Container";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import { NavLink } from "react-router-dom";
import SignUpBox from "./signup/SignUp";
import SearchBox from "./search/Search";

function navbarIconsReducer(state, action) {
  switch (action.type) {
    case "IS_SIGN_OPENED":
      return {
        ...state,
        isSignBoxOpen: !state.isSignBoxOpen,
        isSearchBoxOpen: false,
      };

    case "IS_SEARCH_OPENED":
      return {
        ...state,
        isSearchBoxOpen: !state.isSearchBoxOpen,
        isSignBoxOpen: false,
      };

    case "IS_HAMBURGER_OPENED":
      return {
        isHamburgerOpen: !state.isHamburgerOpen,
        isSignBoxOpen: false,
        isSearchBoxOpen: false,
      };

    default:
      return state;
  }
}

const Navbar = () => {
  const [
    { isSignBoxOpen, isSearchBoxOpen, isHamburgerOpen, isAdminLoggedIn },
    dispatch,
  ] = useReducer(navbarIconsReducer, {
    isSignBoxOpen: false,
    isSearchBoxOpen: false,
    isHamburgerOpen: false,
    isAdminLoggedIn: false,
  });

  const handleStatusOfIcons = (type) => dispatch({ type });

  return (
    <>
      {isSearchBoxOpen && <SearchBox />}
      <Container fluid id="navbar">
        <Row
          className="hamburgerContainer"
          onClick={() => handleStatusOfIcons("IS_HAMBURGER_OPENED")}
        >
          <Col className="hamburgerIcon"></Col>
        </Row>
        <Row
          className={`collapseMenuItems ${
            isHamburgerOpen ? "hamburgerOpened" : ""
          }`}
        >
          <Row className="navbarLogo">LOGO</Row>

          <Row xl={7} lg={7} className="navbarItems">
            <Col>
              <NavLink exact to="/">
                Home
              </NavLink>
            </Col>
            <Col>
              <NavLink to="/about">About</NavLink>
            </Col>
            <Col>
              <NavLink to="/blog">Blog</NavLink>
            </Col>
            <Col>
              <NavLink to="/products">Products</NavLink>
            </Col>
            <Col>
              <NavLink to="/contact">Contact</NavLink>
            </Col>
          </Row>
          <Row xl={2} lg={2} className="icon-trio">
            <div>
              <i
                className="fas fa-search"
                onClick={() => handleStatusOfIcons("IS_SEARCH_OPENED")}
              ></i>
            </div>
            {isAdminLoggedIn ? (
              <div>
                <NavLink to="/admin">
                  <i className="fas fa-user-shield"></i>
                </NavLink>
              </div>
            ) : (
              <div>
                <i
                  className="fas fa-user-circle"
                  onClick={() => handleStatusOfIcons("IS_SIGN_OPENED")}
                ></i>
              </div>
            )}
            <div>
              <NavLink to="/shoppingcart">
                <i className="fas fa-shopping-cart"></i>
              </NavLink>
            </div>
          </Row>
        </Row>
      </Container>
      {isSignBoxOpen && <SignUpBox />}
    </>
  );
};

export default Navbar;
