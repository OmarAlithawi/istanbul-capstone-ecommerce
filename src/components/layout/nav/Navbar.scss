@import "../../common/mixins";
@import "../../common/variables";

@mixin flexCenter {
  display: flex;
  justify-content: center;
  align-items: center;
}
// Fixed Navbar always will be shown over other components except HamburgerIcon, Search and SignUpModal elements.
$navbarIndex: 102;
$hamburgerIconIndex: 103;
$closeSignUpIconIndex: 105;

.navbarWithTransparentBackground {
  background-color: transparent !important;
}

.navbarClickAwayWrapper {
  height: 100px;
  width: 100%;
  position: absolute;
  top: 0;
  overflow: hidden;
}

.shoppingCartLockedSvg {
  background: url("../../../assets/svgs/locked-cart.svg") no-repeat;
  display: inline-block;
  height: 22px;
  width: 25px;
  cursor: pointer;
}

.navbar {
  background-color: $navbarGeneralBlue;
  display: flex;
  height: 100px;
  justify-content: space-around;
  align-items: center;
  color: white;
  justify-content: flex-end;
  position: fixed;
  padding: 0;
  padding-right: 25px;
  overflow: hidden;

  // To show Navbar that has transparent background on top of Header component.
  z-index: $navbarIndex;

  .collapseMenuItems {
    width: 90%;
    justify-content: flex-end;
  }

  .navbarItems {
    width: 50%;
    padding: 10px 5px;

    .navLinkCol {
      padding-left: 5px;
      padding-right: 5px;
      transition: font-size 0.3s ease-in-out;
      @include flexCenter;
    }
  }

  .navbarLogoWrapper {
    position: absolute;
    width: 150px;
    height: 100%;
    @include flexCenter;
    left: -80px;
    top: 50%;
    transform: translateY(-50%);
    cursor: pointer;

    .navbarLogo {
      background: url("../../../assets/svgs/katara_logo.svg") no-repeat;
      color: white;
      width: 100%;
      height: 35px;
    }
  }

  .navbarItemWrapper {
    text-align: center;
    position: relative;

    a {
      text-decoration: none;
      color: white;
    }
  }

  .shoppingCartAccessRejected {
    color: #c13a3a;
    height: 40px;
    font-size: 14px;
  }

  .iconTrio {
    width: 30%;
    position: relative;
    @include flexCenter;
    justify-content: flex-end;
    padding: 10px 30px 10px 5px;

    .shoppingCartIconContainer {
      height: 22px;
    }

    .iconWrapper {
      max-width: 35px;
      margin-right: 12px;
      min-width: 18px;
      @include flexCenter;
    }

    .loggedInUserIconWrapper {
      display: flex;
      flex-wrap: nowrap;
      order: 5;
      background-color: #efe9e9;
      border: 2px solid #8c8c92;
      height: 45px;
      border-radius: 3px;
      color: $navbarGeneralBlue;
      margin: 0;
      padding: 0 10px;
      min-width: 128px !important;

      &:hover {
        cursor: pointer;
      }

      .loggedInTabCol {
        padding: 0;
        flex-wrap: wrap;
        @include flexCenter;

        .loggedDefaultIcon {
          font-size: 19px;
        }

        .loggedArrowIcon {
          font-size: 15px;
        }
      }

      .loggedInTextContainer {
        flex-direction: column;
        flex-wrap: nowrap;
        height: 70%;

        .loggedInTabText {
          margin: 0;
          padding: 0;
          @include flexCenter;
        }

        .loggedProfileText {
          font-weight: 600;
          font-size: 12px;
          line-height: 12px;
        }

        .loggedUserName {
          font-size: 9px;
          line-height: 9px;
        }
      }
    }

    // FontAwesome classNames.
    .fas,
    .fab {
      font-size: 22px;
      cursor: pointer;
      transition: font-size 0.3s ease-in-out;
    }
  }

  .userLoggedIn {
    .iconTrio {
      width: 40%;
    }

    .languageDropdown {
      top: 50px;
    }
  }

  @include media(tablet) {
    .navbarItems {
      margin-right: 20px;
    }

    .navLinkCol {
      font-size: 95%;
    }

    .navbarLogo {
      height: 25px;
      width: 100px;
    }

    .fas,
    .fab {
      font-size: 125%;
    }
  }

  .hamburgerContainer {
    display: none;
    background-color: $navbarGeneralBlue;
  }

  .userLoggedIn .navbarLogo {
    height: 25px;
  }
}

@include media(tablet-sm) {
  .navbar {
    width: 200px;
    height: 100%;
    min-height: 100%;
    right: 0;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    background-color: transparent;
    padding: 0;

    .navbarItems {
      margin: 0;
    }

    .collapseMenuItems {
      align-items: center;
      background-color: #6676a0;
      flex-direction: column;
      justify-content: center;
      height: 100%;
      padding-left: 15px;
      padding-right: 15px;
      width: 200px;
      transform: translateX(200px);
      display: none;
      transition: transform 1s linear;

      &.hamburgerOpened {
        display: flex;
        transform: translateX(0px);
        width: 100%;
      }
    }

    .hamburgerContainer {
      background-color: transparent;
    }

    .hamburgerContainer {
      display: block;
      width: 50px;
      height: 50px;
      position: absolute;
      top: 10px;
      right: 50px;
      z-index: $hamburgerIconIndex;

      .hamburgerIcon {
        display: block;
        width: 60%;
        height: 3px;
        position: absolute;
        left: 50%;
        top: 50%;
        transform: translate(-50%, -50%);
        background-color: white;
        z-index: $hamburgerIconIndex;

        &::before,
        &::after {
          content: "";
          height: 3px;
          position: absolute;
          background-color: white;
          z-index: $hamburgerIconIndex;
        }

        &::before {
          top: -10px;
          width: 50%;
          right: 0;
        }

        &::after {
          bottom: -10px;
          width: 120%;
          right: 0;
        }
      }
    }

    .navbarLogoWrapper {
      height: 40px;
      bottom: 0;
      top: auto;
      left: auto;

      .navbarLogo {
        bottom: 20px;
        height: 25px;
        width: 110px;
      }
    }

    .navbarItems {
      flex-direction: column;
      width: 100%;
      height: 40%;
      order: 2;

      .navLinkCol {
        border-bottom: 0.1px solid #f5deb378;
        transition: all 1.5s ease-in-out;

        .navLinkWrapper {
          @include flexCenter;
        }

        // Add border-top to the first side-nav item to tell them apart.
        &:nth-of-type(1) {
          border-top: 0.1px solid #f5deb378;
        }

        &:hover {
          border-right: 0.1px solid #f5deb378;
          border-left: 0.1px solid #f5deb378;
        }

        a {
          display: block;
          width: 100%;
          height: 100%;
        }
      }
    }

    .iconTrio {
      order: 1;
      width: 100%;
      flex-wrap: nowrap;
      height: 50px;
      margin: 0 auto;
      margin-bottom: 30px;
      padding: 0;
      justify-content: center;

      .languageIconWrapper {
        margin: 0;
      }

      .fas,
      .fab {
        font-size: 22px;
      }
    }

    .userLoggedIn {
      .iconTrio {
        margin: 0;
        width: 100%;
        height: 120px;
        display: flex;
        flex-wrap: wrap;
      }

      .languageDropdown {
        top: 40px;
        left: 35px;
      }
    }
  }
}

// When SearchBox is opened at the top of page, move Navbar down only on fullScreen.
.moveOverHeader {
  transform: translateY(100px);

  @include media(tablet-sm) {
    transform: none;
  }
}

.closeSignUpForm {
  position: fixed !important;
  transform: translateY(70px);
  top: 40px;
  right: 50px;
  z-index: $closeSignUpIconIndex;
  font-size: 25px;
  width: 30px;
  text-align: center;

  &:hover {
    cursor: pointer;
  }

  @include media(tablet-sm) {
    top: calc(50% - 240px);
    right: calc(50% - 150px);
  }
}

@include media(phone-sm) {
  .navbar {
    width: 100%;

    .iconWrapper {
      margin-right: 20px !important;

      .navbarIcon {
        font-size: 30px !important;
      }

      .shoppingCartLockedSvg {
        height: 30px;
        width: 33px;
      }

      .shoppingCartIconContainer {
        height: 30px;

        .shoppingCartIcon {
          font-size: 30px;
        }
      }
    }

    .languageIconWrapper {
      margin: 0 !important;
    }

    .userLoggedIn {
      .languageIconWrapper {
        position: relative;
      }

      .languageDropdown {
        top: 0;
        left: auto;
        min-width: 45px;
        height: 45px;
        right: -50px;
        padding: 0;

        .languageDropdownSelect {
          padding: 5px;
        }
      }

      .iconTrio {
        width: 50%;

        .loggedInUserIconWrapper {
          margin: 0 !important;
          min-width: 150px !important;
        }
      }

      .loggedInTextContainer {
        height: 80% !important;

        .loggedProfileText {
          font-size: 14px !important;
        }

        .loggedUserName {
          font-size: 12px !important;
        }
      }

      .loggedDefaultIcon,
      .loggedArrowIcon {
        font-size: 25px !important;
      }
    }
  }
}
