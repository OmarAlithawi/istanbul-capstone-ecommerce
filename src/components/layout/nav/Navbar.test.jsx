import React from "react";
import { render } from "@testing-library/react";
import Navbar from "./Navbar";
import { BrowserRouter as Router } from "react-router-dom";

it("renders correctly", () => {
  const tree = render(
    <Router>
      <Navbar />
    </Router>
  );
  expect(tree).toMatchSnapshot();
});
