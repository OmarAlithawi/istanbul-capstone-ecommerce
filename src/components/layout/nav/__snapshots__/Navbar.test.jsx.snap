// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
        id="navbar"
      >
        <div
          class="hamburgerContainer row"
        >
          <div
            class="hamburgerIcon col"
          />
        </div>
        <div
          class="collapseMenuItems  row"
        >
          <div
            class="navbarLogo row"
          >
            LOGO
          </div>
          <div
            class="navbarItems row row-cols-xl-7 row-cols-lg-7"
          >
            <div
              class="col"
            >
              <a
                aria-current="page"
                class="active"
                href="/"
              >
                Home
              </a>
            </div>
            <div
              class="col"
            >
              <a
                href="/about"
              >
                About
              </a>
            </div>
            <div
              class="col"
            >
              <a
                href="/blog"
              >
                Blog
              </a>
            </div>
            <div
              class="col"
            >
              <a
                href="/products"
              >
                Products
              </a>
            </div>
            <div
              class="col"
            >
              <a
                href="/contact"
              >
                Contact
              </a>
            </div>
          </div>
          <div
            class="icon-trio row row-cols-xl-2 row-cols-lg-2"
          >
            <div>
              <i
                class="fas fa-search"
              />
            </div>
            <div>
              <i
                class="fas fa-user-circle"
              />
            </div>
            <div>
              <a
                href="/shoppingcart"
              >
                <i
                  class="fas fa-shopping-cart"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
      id="navbar"
    >
      <div
        class="hamburgerContainer row"
      >
        <div
          class="hamburgerIcon col"
        />
      </div>
      <div
        class="collapseMenuItems  row"
      >
        <div
          class="navbarLogo row"
        >
          LOGO
        </div>
        <div
          class="navbarItems row row-cols-xl-7 row-cols-lg-7"
        >
          <div
            class="col"
          >
            <a
              aria-current="page"
              class="active"
              href="/"
            >
              Home
            </a>
          </div>
          <div
            class="col"
          >
            <a
              href="/about"
            >
              About
            </a>
          </div>
          <div
            class="col"
          >
            <a
              href="/blog"
            >
              Blog
            </a>
          </div>
          <div
            class="col"
          >
            <a
              href="/products"
            >
              Products
            </a>
          </div>
          <div
            class="col"
          >
            <a
              href="/contact"
            >
              Contact
            </a>
          </div>
        </div>
        <div
          class="icon-trio row row-cols-xl-2 row-cols-lg-2"
        >
          <div>
            <i
              class="fas fa-search"
            />
          </div>
          <div>
            <i
              class="fas fa-user-circle"
            />
          </div>
          <div>
            <a
              href="/shoppingcart"
            >
              <i
                class="fas fa-shopping-cart"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
